### ISSUE TITLE
RuntimeException in version 1.1 jar in Burp 2.1

### ISSUE THREAD

**Issue by floyd-fuh** [CoreyD97/Stepper#3]
Hi Corey,

I'm able to load the version 1.1 jar in Burp 2.1 but the Error tab will show the following:

```
java.lang.RuntimeException: java.lang.RuntimeException: org.python.core.PyException
	at burp.ccx.a(Unknown Source)
	at burp.b89.<init>(Unknown Source)
	at burp.f7g.a(Unknown Source)
	at burp.b91.b(Unknown Source)
	at burp.b91.a(Unknown Source)
	at burp.fbn.a(Unknown Source)
	at burp.gby.registerMessageEditorTabFactory(Unknown Source)
	at burp.ge3.registerMessageEditorTabFactory(Unknown Source)
	at burp.a7u.registerMessageEditorTabFactory(Unknown Source)
	at com.coreyd97.stepper.Stepper.lambda$registerExtenderCallbacks$0(Stepper.java:66)
	at java.desktop/java.awt.event.InvocationEvent.dispatch(InvocationEvent.java:313)
	at java.desktop/java.awt.EventQueue.dispatchEventImpl(EventQueue.java:770)
	at java.desktop/java.awt.EventQueue$4.run(EventQueue.java:721)
	at java.desktop/java.awt.EventQueue$4.run(EventQueue.java:715)
	at java.base/java.security.AccessController.doPrivileged(Native Method)
	at java.base/java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(ProtectionDomain.java:85)
	at java.desktop/java.awt.EventQueue.dispatchEvent(EventQueue.java:740)
	at java.desktop/java.awt.EventDispatchThread.pumpOneEventForFilters(EventDispatchThread.java:203)
	at java.desktop/java.awt.EventDispatchThread.pumpEventsForFilter(EventDispatchThread.java:124)
	at java.desktop/java.awt.EventDispatchThread.pumpEventsForHierarchy(EventDispatchThread.java:113)
	at java.desktop/java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:109)
	at java.desktop/java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:101)
	at java.desktop/java.awt.EventDispatchThread.run(EventDispatchThread.java:90)
java.lang.RuntimeException: java.lang.RuntimeException: org.python.core.PyException
	at burp.ccx.a(Unknown Source)
	at burp.b89.<init>(Unknown Source)
	at burp.f7g.a(Unknown Source)
	at burp.b91.b(Unknown Source)
	at burp.b91.a(Unknown Source)
	at burp.fbn.a(Unknown Source)
	at burp.gby.registerMessageEditorTabFactory(Unknown Source)
	at burp.ge3.registerMessageEditorTabFactory(Unknown Source)
	at burp.a7u.registerMessageEditorTabFactory(Unknown Source)
	at com.coreyd97.stepper.Stepper.lambda$registerExtenderCallbacks$0(Stepper.java:66)
	at java.desktop/java.awt.event.InvocationEvent.dispatch(InvocationEvent.java:313)
	at java.desktop/java.awt.EventQueue.dispatchEventImpl(EventQueue.java:770)
	at java.desktop/java.awt.EventQueue$4.run(EventQueue.java:721)
	at java.desktop/java.awt.EventQueue$4.run(EventQueue.java:715)
	at java.base/java.security.AccessController.doPrivileged(Native Method)
	at java.base/java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(ProtectionDomain.java:85)
	at java.desktop/java.awt.EventQueue.dispatchEvent(EventQueue.java:740)
	at java.desktop/java.awt.EventDispatchThread.pumpOneEventForFilters(EventDispatchThread.java:203)
	at java.desktop/java.awt.EventDispatchThread.pumpEventsForFilter(EventDispatchThread.java:124)
	at java.desktop/java.awt.EventDispatchThread.pumpEventsForHierarchy(EventDispatchThread.java:113)
	at java.desktop/java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:109)
	at java.desktop/java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:101)
	at java.desktop/java.awt.EventDispatchThread.run(EventDispatchThread.java:90)
```

The Python exception is very weird, especially as your extension is written in Java (I'm also loading it as a Java extension).

Off topic, but have you submitted this extension to be included in the BApp store?

**Comment by CoreyD97** [comment]
Hey, 
That's an odd issue as like you say. The extension is written in Java and the exception comes from further down in the stack than Stepper itself. 

Do you have the same issue if you load Stepper on its own without any other extensions?

As for the BApp store, I requested it a while ago but have been too busy to chase it recently. I might try chasing it up soon.

**Comment by floyd-fuh** [comment]
Ok, I just checked and this is only happending when my UploadScanner is loaded at the same time... which makes absolutely no sense to me. I'll report it to the Portswigger guys and will ask for a deobfuscation of the stack trace.

**Comment by floyd-fuh** [comment]
Reported at https://support.portswigger.net/customer/en/portal/questions/17597059-stack-trace-when-loading-extensions?new=17597059

**Comment by floyd-fuh** [comment]
So I think this is a problem in the Stepper extension, because it doesn't happen as long as it is not loaded, but as soon as it is loaded, I get errors in a lot of extensions. Is it possible that the Stepper extension passes an HTTPService of null? Because in the Logger++ Error tab I see:

```
java.lang.IllegalArgumentException: HTTP service cannot be null
	at burp.dtj.ai(Unknown Source)
	at burp.g0d.getHttpService(Unknown Source)
	at burp.brl.getHttpService(Unknown Source)
	at com.coreyd97.stepper.ui.VariableReplacementsTabFactory.findActualController(VariableReplacementsTabFactory.java:35)
	at com.coreyd97.stepper.ui.VariableReplacementsTabFactory.createNewInstance(VariableReplacementsTabFactory.java:24)
	at burp.b89.<init>(Unknown Source)
	at burp.f7g.a(Unknown Source)
	at burp.b91.b(Unknown Source)
	at burp.b91.a(Unknown Source)
	at burp.fbn.a(Unknown Source)
	at burp.gby.registerMessageEditorTabFactory(Unknown Source)
	at burp.ge3.registerMessageEditorTabFactory(Unknown Source)
	at burp.a7u.registerMessageEditorTabFactory(Unknown Source)
	at com.coreyd97.stepper.Stepper.lambda$registerExtenderCallbacks$0(Stepper.java:66)
	at java.desktop/java.awt.event.InvocationEvent.dispatch(InvocationEvent.java:313)
	at java.desktop/java.awt.EventQueue.dispatchEventImpl(EventQueue.java:770)
	at java.desktop/java.awt.EventQueue$4.run(EventQueue.java:721)
	at java.desktop/java.awt.EventQueue$4.run(EventQueue.java:715)
	at java.base/java.security.AccessController.doPrivileged(Native Method)
	at java.base/java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(ProtectionDomain.java:85)
	at java.desktop/java.awt.EventQueue.dispatchEvent(EventQueue.java:740)
	at java.desktop/java.awt.EventDispatchThread.pumpOneEventForFilters(EventDispatchThread.java:203)
	at java.desktop/java.awt.EventDispatchThread.pumpEventsForFilter(EventDispatchThread.java:124)
	at java.desktop/java.awt.EventDispatchThread.pumpEventsForHierarchy(EventDispatchThread.java:113)
	at java.desktop/java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:109)
	at java.desktop/java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:101)
	at java.desktop/java.awt.EventDispatchThread.run(EventDispatchThread.java:90)
```

**Comment by CoreyD97** [comment]
Ah! Nice catch! I see what the issue is now. 

The Burp API tells message editors who the class that controls them is. However it has a "proxy class" that sits between the true controller and the message editor. This stops us being able to call our own methods in the controller. Which we need to do to get our variables etc. 

To get around this, the message editor can ask the controller for the http service which it's targeting, so I had it initialise with a random value. The message editor then asks each controller what it's httpservice is until it finds one with a matching value to find its true controller (so we don't have to use the proxy controller).

However, once it's found it's controller, it sets its httpservice to null. This part is what is causing your issue! 

Ill tweak that so it doesn't end up with a null value and push a new release. See if that solves your issue.

**Comment by CoreyD97** [comment]
Hey! I wasn't able to reproduce this issue and looking at the code can't find any reason for the initial error to arise.

The IllegalArgumentException was caused when an IMessageEditorController didn't yet have a HttpService set so when Stepper tried to match up the controller object with the step controller, it failed. I've added a catch for this case to stop this arising in future.

Do you have an example of another extension which causes the initial error to be thrown when stepper is also in use? If so, I'll be able to better track the cause of this.

Thanks,
Corey

**Comment by floyd-fuh** [comment]
Hi Corey,

sorry for the late reply. I just installed version 1.2 of Stepper in Burp Pro v2.1.03 and I still have the same issue. I can reproduce by opening a new temporary project, loading the Logger++ extension and then loading Stepper 1.2:

```
java.lang.IllegalArgumentException: HTTP service cannot be null
	at burp.f3g.a(Unknown Source)
	at burp.en1.getHttpService(Unknown Source)
	at burp.b1o.getHttpService(Unknown Source)
	at com.coreyd97.stepper.ui.VariableReplacementsTabFactory.findActualController(VariableReplacementsTabFactory.java:35)
	at com.coreyd97.stepper.ui.VariableReplacementsTabFactory.createNewInstance(VariableReplacementsTabFactory.java:24)
	at burp.b47.<init>(Unknown Source)
	at burp.a31.a(Unknown Source)
	at burp.hs.a(Unknown Source)
	at burp.co6.b(Unknown Source)
	at burp.a31.addNotify(Unknown Source)
	at java.desktop/java.awt.Container.addNotify(Container.java:2800)
	at java.desktop/javax.swing.JComponent.addNotify(JComponent.java:4783)
	at java.desktop/java.awt.Container.addNotify(Container.java:2800)
	at java.desktop/javax.swing.JComponent.addNotify(JComponent.java:4783)
	at java.desktop/java.awt.Container.addNotify(Container.java:2800)
	at java.desktop/javax.swing.JComponent.addNotify(JComponent.java:4783)
	at java.desktop/java.awt.Container.addNotify(Container.java:2800)
	at java.desktop/javax.swing.JComponent.addNotify(JComponent.java:4783)
	at java.desktop/java.awt.Container.addNotify(Container.java:2800)
	at java.desktop/javax.swing.JComponent.addNotify(JComponent.java:4783)
	at java.desktop/java.awt.Container.addNotify(Container.java:2800)
	at java.desktop/javax.swing.JComponent.addNotify(JComponent.java:4783)
	at java.desktop/java.awt.Container.addNotify(Container.java:2800)
	at java.desktop/javax.swing.JComponent.addNotify(JComponent.java:4783)
	at java.desktop/java.awt.Container.addImpl(Container.java:1146)
	at java.desktop/javax.swing.JTabbedPane.insertTab(JTabbedPane.java:730)
	at com.coreyd97.stepper.ui.StepperUI.onStepSequenceAdded(StepperUI.java:63)
	at com.coreyd97.stepper.Stepper.addStepSequence(Stepper.java:104)
	at com.coreyd97.stepper.StateManager.loadSavedSequences(StateManager.java:26)
	at com.coreyd97.stepper.Stepper.lambda$registerExtenderCallbacks$0(Stepper.java:70)
	at java.desktop/java.awt.event.InvocationEvent.dispatch(InvocationEvent.java:313)
	at java.desktop/java.awt.EventQueue.dispatchEventImpl(EventQueue.java:770)
	at java.desktop/java.awt.EventQueue$4.run(EventQueue.java:721)
	at java.desktop/java.awt.EventQueue$4.run(EventQueue.java:715)
	at java.base/java.security.AccessController.doPrivileged(Native Method)
	at java.base/java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(ProtectionDomain.java:85)
	at java.desktop/java.awt.EventQueue.dispatchEvent(EventQueue.java:740)
	at java.desktop/java.awt.EventDispatchThread.pumpOneEventForFilters(EventDispatchThread.java:203)
	at java.desktop/java.awt.EventDispatchThread.pumpEventsForFilter(EventDispatchThread.java:124)
	at java.desktop/java.awt.EventDispatchThread.pumpEventsForHierarchy(EventDispatchThread.java:113)
	at java.desktop/java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:109)
	at java.desktop/java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:101)
	at java.desktop/java.awt.EventDispatchThread.run(EventDispatchThread.java:90)
```

However, I'm unsure if it is still exactly the same issue as before, but it is again related to the missing HttpService.

Another extension that triggers the issue as well is my UploadScanner extension.

**Comment by CoreyD97** [comment]
Hey Floyd,

Thanks for this, I'll take another look with that info. It's quite odd though as stepper initialises the message editors with a HTTP service value and I couldn't find any way which could result in a null value being set when I looked at this issue previously.

It's also quite odd that other extensions which register message editors trigger the exception, as Stepper only checks the messageEditorControllers which were registered by itself, and actually has a try catch around the line which is throwing the illegalargumentexception too. 

Does the extension still function properly when this exception is thrown? Let me know if you see any odd behaviour after this error. 

Thanks,
Corey

**Comment by floyd-fuh** [comment]
I'm sometimes not sure if a try-catch around Burp API really helps, because you might be passing a invalid object to the Burp API (which it should process) and it catches it and prints it to the error console of the extension but might not allow you to catch it. Very unsure though.

From a functionality point of view there is nothing to worry I think. I'm just a perfectionist sometimes and I've seen many many many other Burp extension randomly fail, throw exceptions and not do what they are supposed to do. But even more often extensions are just a performance nightmare. I guess I'm the only person who wrote an extension that asks users to report crashes (https://github.com/modzero/mod0BurpUploadScanner/blob/master/UploadScanner.py#L747). Not sure if this is crazy from my side though. But I don't think Stepper is one of those extensions that fails, it works very well in my experience. Actually I think it would be nice if Portswigger would provide sequence and regex-variable UI elements, so other extensions could simply define the behavior of the important request (e.g. an injection) and at the same time get the contents of all the sequence request/responses. It would be good to have it in UploadScanner. So again, good work on this extension :)

**Comment by CoreyD97** [comment]
This case is slightly different in regards to the try catch, as it is around the getter method. Burp logs the error before throwing the exception as a warning to developers not to set httpservice to null and to fix their extension but in this case I can't see why it's getting a null value at all there. The catch block here just makes sure the extension falls back to functional, but less preferable behaviour if it does fail.

I agree with the perfectionist angle though, I know for sure there's lots of bugs in my extensions and I'm slowly working to fix them thanks to people like you going out of your way to report them, hopefully none of them are fatal crashes :grinning:

It would definitely be useful to have some extra objects for common functionality, I'm actually trying to implement my own little library of them as I come to need them in my BurpExtenderUtilities project. At the minute it's basic, but has a preferences store and basic UI building helpers and some UI elements like the history list I use in Logger++, pop out panel and I've got some plans to add quite a few more things to it in future.

If you find you're needing anything or have developed anything which could be useful as a common object feel free to submit a PR or issue to the project, I'd be happy to include it!

**Comment by CoreyD97** [comment]
Hey Floyd,

So after doing some digging, it seems that some extensions don't set values for the HTTP service when implementing IMessageEditorController, and it seems Burp itself potentially might not be setting this value until later on, after all tabs have been added.

Therefore, when an IMessageEditor was created, stepper was looking to see if it was a message editor created by stepper itself to be able to match the message editor to the step controller however for editors which did not yet have a HTTP service set, Burp was throwing an error.

I have modified the way Stepper matches the editor to its controller, and instead of using the HTTP service, Stepper sets a random request body which is then removed once the match has been completed. This will stop the message from being displayed and actually results in a much smoother method of matching. I don't know why I didn't think of this earlier!

Thanks for your help in identifying this issue!
- Corey
